core/src/io/anuke/mindustry/world/consumers/Consumers.java

```diff
@@ -87,6 +121,15 @@ public boolean has(Class<? extends Consume> type){
         return map.containsKey(type);
     }
 
+    public boolean hasSubtypeOf(Class<? extends Consume> type){
```

Comment:
I would highly recommend avoiding class reflection (`isAssignableFrom()`) and O(n) search in this method. If a block has a lot of consumers, it's going to need to search through every single one for every tile in the graph.

All this searching is here to accommodate different types of power consumers, and I don't think it's worth it. The only class that extends `ConsumePower` is `ConsumeForceProjectorPower`, and there are ways to get around this.

A quick fix would involve changing:
```java
    public Consume add(Consume consume){
        map.put(consume.getClass(), consume);
        return consume;
    }
```
to:
```java
    public Consume add(Consume consume){
        map.put((consume instanceof ConsumePower ? ConsumePower.class : consume.getClass()), consume);
        return consume;
    }
```

This would ensure that all types of power consumers are registered under the same class type (ConsumePower).

Reply:
To be honest, I don't think there should be searching at all. Just create an `enum` for the possible types of consumers (6 at most) and use a regular array to store the consumers, `null` for the non-existent ones. The consumers which do basically the same thing (and have the same interface) but have different implementations could be stored in the same field of the array.

Reply:
Using classes as keys in the map has the benefit of automatically casting to the right type, e.g. `get(ConsumeItem.class) -> returns ConsumeItem`. But I don't suppose it matters too much, since you could still do `<ConsumeItem>get(ConsumeType.item)` or just use the utility methods for that purpose. Performance would be equivalent; it's a matter of design.

Reply:
I would suggest doing the quick fix for now and implementing the enum thing a new branch if desired. The hasSubtypeOf was introduced since tests failed for the force projector, but doing what you suggested in the fix would be a cleaner and more efficient solution indeed.